### ISSUE TITLE
Lowering of disk watermark should trigger index allocation

### ISSUE THREAD

**Issue by kkewwei** [elastic/elasticsearch#99333]
### Elasticsearch Version

7.10.2

### Java Version

openjdk version "13.0.2" 2020-01-14

### OS Version

Darwin DKXN6FXCH7 21.5.0 Darwin Kernel Version 21.5.0

### Problem Description

In the cluster, disk watermark is high than the high threshold, and the new index1 is created, as it's known, all shards of the index1 will be not assigned. The usage of disk decreased for some reason, now the disk usage is lower than high watermark, higher than low watermark, but the primary shards of index1 are still not be assigned, this is not normal. the result of api `_cluster/allocation/explain` is as follows:
```
{
   "index": "index1",
   "shard": 0,
   "primary": true,
   "current_state": "unassigned",
   "unassigned_info": {
      "reason": "INDEX_CREATED",
      "at": "2023-09-08T03:52:14.851Z",
      "last_allocation_status": "no"
   },
   "can_allocate": "yes",
   "allocate_explanation": "Elasticsearch can allocate the shard.",
   "target_node": {
      "id": "VoMSBNvQTyuxjSndDoG0qg",
      "name": "node2",
      "transport_address": "127.0.0.1:9302",
      "attributes": {
         "xpack.installed": "true"
      }
   },
   "node_allocation_decisions": [
      {
         "node_id": "VoMSBNvQTyuxjSndDoG0qg",
         "node_name": "node2",
         "transport_address": "127.0.0.1:9302",
         "node_attributes": {
            "xpack.installed": "true"
         },
         "node_decision": "yes",
         "weight_ranking": 1
      }
......
```
the explain show that "Elasticsearch can allocate the shard, but shard can't be assigned"

We only call `reroute` when the disk usage is lower than low watermark, we should also call `reroute` when the disk watermark is lower than high watermark or higher than the low watermark.
https://github.com/elastic/elasticsearch/blob/e8f8a0ac2eee36588e97c9a4eed9f64e25f33f78/server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java#L260

### Steps to Reproduce

Current disk usage is 50%.
1. Create the file1 which size is 10Gb.
2. set the setting `cluster.routing.allocation.disk.watermark.high` to be 49.999%
3. create the index1.
4. delete file1. the disk usage will lower than 49.999%

The index1 will be not assigned.

### Logs (if relevant)

[2023-09-03T13:56:40,738+0800][INFO][o.e.c.r.a.AllocationService] [master1] Cluster health status changed from [YELLOW] to [RED] (reason: [index [vecompass-log-time-2023.09.03-000014] created]).
[2023-09-03T13:56:40,773+0800][INFO][c.a.o.i.i.ManagedIndexCoordinator] [master1] Index [vecompass-log-time-2023.09.03-000014] will be managed by policy [vestack-log-ism]
[2023-09-03T13:57:12,303+0800][WARN][o.e.c.r.a.DiskThresholdMonitor] [master1] high disk watermark [90%] exceeded on [uslvEQqiSc248VPAPQclEg][master0][/es-data/nodes/0] free: 8.4gb[8.5%], shards will be relocated away from this node; currently relocating away shards totalling [0] bytes; the node is expected to continue to exceed the high disk watermark when these relocations are complete
......
[2023-09-03T14:01:12,507+0800][WARN][o.e.c.r.a.DiskThresholdMonitor] [master1] high disk watermark [90%] exceeded on [m0Sid36tSF-tD_5mC0X8eQ][master2][/es-data/nodes/0] free: 5gb[5.1%], shards will be relocated away from this node; currently relocating away shards totalling [0] bytes; the node is expected to continue to exceed the high disk watermark when these relocations are complete
[2023-09-03T14:01:12,507+0800][WARN][o.e.c.r.a.DiskThresholdMonitor] [master1] high disk watermark [90%] exceeded on [NTM1PxiiQByNGHa13jvJTw][master1][/es-data/nodes/0] free: 7.4gb[7.6%], shards will be relocated away from this node; currently relocating away shards totalling [0] bytes; the node is expected to continue to exceed the high disk watermark when these relocations are complete
[2023-09-03T14:02:12,565+0800][INFO][o.e.c.r.a.DiskThresholdMonitor] [master1] high disk watermark [90%] no longer exceeded on [uslvEQqiSc248VPAPQclEg][master0][/es-data/nodes/0] free: 14.5gb[14.8%], but low disk watermark [85%] is still exceeded
[2023-09-03T14:02:12,565+0800][INFO][o.e.c.r.a.DiskThresholdMonitor] [master1] high disk watermark [90%] no longer exceeded on [m0Sid36tSF-tD_5mC0X8eQ][master2][/es-data/nodes/0] free: 11.4gb[11.7%], but low disk watermark [85%] is still exceeded
[2023-09-03T14:02:12,565+0800][INFO][o.e.c.r.a.DiskThresholdMonitor] [master1] high disk watermark [90%] no longer exceeded on [NTM1PxiiQByNGHa13jvJTw][master1][/es-data/nodes/0] free: 13.7gb[14%], but low disk watermark [85%] is still exceeded
......
[2023-09-03T15:44:39,081+0800][INFO][o.e.c.s.ClusterSettings] [master1] updating [cluster.routing.allocation.node_concurrent_incoming_recoveries] from [8] to [9]
[2023-09-03T15:44:39,259+0800][INFO][o.e.c.r.a.AllocationService] [master1] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[vecompass-log-time-2023.09.03-000014][7]]]).
......
[2023-09-03T15:46:48,354+0800][INFO][o.e.c.r.a.DiskThresholdMonitor] [master1] low disk watermark [85%] no longer exceeded on [uslvEQqiSc248VPAPQclEg][master0][/es-data/nodes/0] free: 15.8gb[16.1%]
[2023-09-03T15:46:48,354+0800][INFO][o.e.c.r.a.DiskThresholdMonitor] [master1] low disk watermark [85%] no longer exceeded on [NTM1PxiiQByNGHa13jvJTw][master1][/es-data/nodes/0] free: 14.6gb[15%]
[2023-09-03T15:47:18,416+0800][INFO][o.e.c.r.a.DiskThresholdMonitor] [master1] low disk watermark [85%] exceeded on [NTM1PxiiQByNGHa13jvJTw][master1][/es-data/nodes/0] free: 14.3gb[14.6%], replicas will not be assigned to this node

To assign the shard, i change the `cluster.routing.allocation.node_concurrent_incoming_recoveries` to trigger the `reroute`.

**Comment by gbanasiak** [comment]
@kkewwei I did a quick test in recent 8.9.x and couldn't reproduce this behaviour. Are you able to reproduce in 7.17.x or 8.x instead of 7.10.2 (see https://www.elastic.co/support/eol)?

**Comment by kkewwei** [comment]
@gbanasiak I test in 8.8.1, it indeed also exists. If you delete by rm file in linux, not delete index?

**Comment by gbanasiak** [comment]
Thank you for checking. My test consisted of lowering the watermark level artificially through `PUT _cluster/settings` instead of adding/removing a file on disk. I've labeled the issue for the appropriate team to investigate.

**Comment by DaveCTurner** [comment]
> My test consisted of lowering the watermark level artificially through PUT _cluster/settings instead of adding/removing a file on disk.

Adjusting cluster settings will always trigger a reroute, so I don't think you can see the effect described in the OP this way.

**Comment by kkewwei** [comment]
@DaveCTurner yes, to lower the disk watermark in reproduce, I have to delete file to avoid `reroute` instead of deleting the index. It does happen in our product. 

@DaveCTurner @gbanasiak if it's a bug, I'm please to fix it.
